File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int len = 0;
        int li = 0;
        int ri = 0;
        
        if(s.size() < 2) return s.size();
        
        unordered_map<char, int> counter;
        
        while(ri < s.size()){
            if(counter.find(s[ri]) == counter.end()){
                counter[s[ri]]++;
                ri++;
            }else{
                counter[s[li]]--;
                if(counter[s[li]] == 0) counter.erase(s[li]);
                li++;
            }
            
            len = max(len, ri-li);
        }
        
        return len;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s)==0:
            return 0
        l, r = 0, 0
        checker = set()
        res = 0
        
        while r < len(s):
            while s[r] in checker:
                checker.remove(s[l])
                l += 1
                
            checker.add(s[r])
            res = max(res, r-l+1)
            r += 1
                
        return res
        